modules/stochastic_tools/include/utils/Calculators.h

```diff
@@ -43,67 +47,117 @@ MultiMooseEnum makeCalculatorEnum();
 };
 
-class StdDev : public Calculator
+template <typename InType, typename OutType>
+class StdDev : public Calculator<InType, OutType>
 {
 public:
-  StdDev(const libMesh::ParallelObject &);
-  virtual Real compute(const std::vector<Real> &, bool) const override;
+  using Calculator<InType, OutType>::Calculator;
+  virtual OutType compute(const InType &, bool) const override;
 };
 
-class StdErr : public StdDev
+template <typename InType, typename OutType>
+class StdErr : public StdDev<InType, OutType>
 {
 public:
-  StdErr(const libMesh::ParallelObject &);
-  virtual Real compute(const std::vector<Real> &, bool) const override;
+  using StdDev<InType, OutType>::StdDev;
+  virtual OutType compute(const InType &, bool) const override;
 };
 
-class Ratio : public Calculator
+template <typename InType, typename OutType>
+class Ratio : public Calculator<InType, OutType>
 {
 public:
-  Ratio(const libMesh::ParallelObject &);
-  virtual Real compute(const std::vector<Real> &, bool) const override;
+  using Calculator<InType, OutType>::Calculator;
+  virtual OutType compute(const InType &, bool) const override;
 };
 
-class L2Norm : public Calculator
+template <typename InType, typename OutType>
+class L2Norm : public Calculator<InType, OutType>
 {
 public:
-  L2Norm(const libMesh::ParallelObject &);
-  virtual Real compute(const std::vector<Real> &, bool) const override;
+  using Calculator<InType, OutType>::Calculator;
+  virtual OutType compute(const InType &, bool) const override;
 };
+
+/*
+ * Free function for building a const Calculator object for use by Statistics object
+ */
+std::unique_ptr<const Calculator<std::vector<Real>, Real>>
+makeCalculator(const MooseEnumItem & item, const libMesh::ParallelObject & other);
+
+/*
+template <typename InType, typename OutType>
+std::unique_ptr<const Calculator<InType, OutType>> makeCalculator(const MooseEnumItem & item,
+                                                                  const libMesh::ParallelObject &
+other)
+{
+  if (item == "min")
+    return libmesh_make_unique<const Min<InType, OutType>>(other, item);
+
+  else if (item == "max")
+    return libmesh_make_unique<const Max<InType, OutType>>(other, item);
+
+  else if (item == "sum")
+    return libmesh_make_unique<const Sum<InType, OutType>>(other, item);
+
+  else if (item == "mean" || item == "average") // average is deprecated
+    return libmesh_make_unique<const Mean<InType, OutType>>(other, item);
+
+  else if (item == "stddev")
+    return libmesh_make_unique<const StdDev<InType, OutType>>(other, item);
+
+  else if (item == "stderr")
+    return libmesh_make_unique<const StdErr<InType, OutType>>(other, item);
+
+  else if (item == "norm2")
+    return libmesh_make_unique<const L2Norm<InType, OutType>>(other, item);
+
+  else if (item == "ratio")
+    return libmesh_make_unique<const Ratio<InType, OutType>>(other, item);
+
+  ::mooseError("Failed to create Statistics::Calculator object for ", item);
+  return nullptr;
+}
+*/
```

Comment:
@loganharbour When I try to use this template function, the compiler no like.

Reply:
Do you have a compiler error?